futures/src/channel.rs

```diff
@@ -384,46 +459,56 @@ impl<T: AsyncRead+AsyncWrite+Send+'static> Channel<T> {
+            transport.conn.channel_close(channel_id, code, message, 0, 0).map(|_| None)
         })
     }
 
     /// ack a channel close
-    pub fn close_ok(&self) -> Box<Future<Item = (), Error = io::Error> + Send + 'static> {
-        self.run_on_locked_transport("close_ok", "Could not ack closed channel", |transport| {
-            transport.conn.channel_close_ok(self.id).map(|_| None)
+    pub fn close_ok(&self) -> impl Future<Item = (), Error = io::Error> + Send {
+        let channel_id = self.id;
+        self.run_on_locked_transport("close_ok", "Could not ack closed channel", move |transport| {
+            transport.conn.channel_close_ok(channel_id).map(|_| None)
         })
     }
 
     /// update a channel flow
-    pub fn channel_flow(&self, options: &ChannelFlowOptions) -> Box<Future<Item=(), Error=io::Error> + Send + 'static> {
-        self.run_on_locked_transport("channel_flow", "Could not update channel flow", |transport| {
-            transport.conn.channel_flow(self.id, options.active).map(|_| None)
+    pub fn channel_flow(&self, options: &ChannelFlowOptions) -> impl Future<Item = (), Error = io::Error> + Send {
+        let channel_id = self.id;
+        let options_active = options.active;
+        self.run_on_locked_transport("channel_flow", "Could not update channel flow", move |transport| {
+            transport.conn.channel_flow(channel_id, options_active).map(|_| None)
         })
     }
 
     /// ack an update to a channel flow
-    pub fn channel_flow_ok(&self, options: &ChannelFlowOptions) -> Box<Future<Item=(), Error=io::Error> + Send + 'static> {
-        self.run_on_locked_transport("channel_flow_ok", "Could not ack update to channel flow", |transport| {
-            transport.conn.channel_flow_ok(self.id, options.active).map(|_| None)
+    pub fn channel_flow_ok(&self, options: &ChannelFlowOptions) -> impl Future<Item = (), Error = io::Error> + Send {
+        let channel_id = self.id;
+        let options_active = options.active;
+        self.run_on_locked_transport("channel_flow_ok", "Could not ack update to channel flow", move |transport| {
+            transport.conn.channel_flow_ok(channel_id, options_active).map(|_| None)
         })
     }
 
-    fn run_on_locked_transport_full<Action, Finished>(&self, method: &str, error: &str, action: Action, finished: Finished, payload: Option<(u16, &[u8], BasicProperties)>) -> Box<Future<Item = Option<bool>, Error = io::Error> + Send + 'static>
-        where Action:   Fn(&mut AMQPTransport<T>) -> Result<Option<RequestId>, lapin_async::error::Error>,
+    fn run_on_locked_transport_full<Action, Finished>(&self, method: &str, error: &str, action: Action, finished: Finished, payload: Option<(u16, &[u8], BasicProperties)>) -> impl Future<Item = Option<bool>, Error = io::Error> + Send
+        where Action:   FnOnce(&mut AMQPTransport<T>) -> Result<Option<RequestId>, lapin_async::error::Error>,
```

Comment:
Why this FnOnce change?

Reply:
Because it doesn't compile otherwise. If you believe that there surely must be a `Fn`, you could try to figure out the signature yourself (I failed personally).

Reply:
What did the compilation error look like with Fn?
It seems weird to me that impl trait causes such a regression, and if it does, I'd prefer not to use it there.